x-pack/packages/ml/response_stream/client/fetch_stream.ts

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import startsWith from 'lodash/startsWith';
+import type { Reducer, ReducerAction } from 'react';
+
+import type { HttpSetup } from '@kbn/core/public';
+
+type GeneratorError = string | null;
+
+/**
+ * Uses `fetch` and `getReader` to receive an API call as a stream with multiple chunks
+ * as soon as they are available. `fetchStream` is implemented as a generator that will
+ * yield/emit chunks and can be consumed for example like this:
+ *
+ * ```js
+ * for await (const [error, chunk] of fetchStream(...) {
+ *     ...
+ * }
+ * ```
+ *
+ * @param endpoint     — The API endpoint including the Kibana basepath.
+ * @param apiVersion   - Optional API version to be used.
+ * @param abortCtrl    — Abort controller for cancelling the request.
+ * @param body         — The request body. For now all requests are POST.
+ * @param ndjson       — Boolean flag to receive the stream as a raw string or NDJSON.
+ *
+ * @returns            - Yields/emits items in the format [error, value]
+ *                       inspired by node's recommended error convention for callbacks.
+ */
+export async function* fetchStream<B extends object, R extends Reducer<any, any>>(
```

Comment:
Played around with this but not sure it's possible: 
- Tried just `R extends Reducer` but that doesn't work because React expects the generics to be set and doesn't provide defaults.
- React uses the same pattern in its type definitions so it actually is `Reducer<any, any>` we should be extending from, anything else would not be the same as it's defined in React itself.

Examples from React definitions:

```ts
    type ReducerState<R extends Reducer<any, any>> = R extends Reducer<infer S, any> ? S : never;
    type ReducerAction<R extends Reducer<any, any>> = R extends Reducer<any, infer A> ? A : never;
    function useReducer<R extends Reducer<any, any>, I>(
        reducer: R,
        initializerArg: I & ReducerState<R>,
        initializer: (arg: I & ReducerState<R>) => ReducerState<R>
    ): [ReducerState<R>, Dispatch<ReducerAction<R>>];
```